foundry/foundry.py

```diff
@@ -365,12 +392,201 @@ def publish(self, foundry_metadata, data_source, title, authors, update=False,
+        https_base_url = endpoint['https_server']
+
+        # Submit data (folders of files or an independent file) to be written to endpoint
+        if os.path.isdir(local_data_path):
+            self._upload_folder(local_data_path, https_base_url, dest_path, endpoint_id)
+        elif os.path.isfile(local_data_path):
+            self._upload_file(local_data_path, https_base_url, dest_path, endpoint_id)
+        else:
+            raise IOError(f"Data path '{local_data_path}' is of unknown type")
+
+        # return the data source URL for publication to MDF
+        return self.make_globus_link(endpoint_id, dest_path)
+
+    def _upload_folder(self, local_data_path, https_base_url, parent_dest_path, endpoint_id):
+        """Upload a folder to a Globus endpoint using HTTPS
+        Args:
+            local_data_path (str): The path to the local data to upload. Can be relative or absolute.
+            https_base_url (str): The URL for a given Globus endpoint.
+            parent_dest_path (str): The path to the parent folder to be written to on the given endpoint. The contents
+                of 'local_data_path' will be written here, including subdirectories.
+            endpoint_id (str): The UUID designating the exact Globus endpoint. Can be obtained via the Globus Web UI or
+                the SDK. This must be the same endpoint pointed to by the https_base_url.
+        Returns
+        -------
+            (list): A list of all of the HTTPS PUT request results (dicts) from the uploads
+        """
+        results = []
+        # initialize destination path as the parent destination path
+        dest_path = parent_dest_path
+
+        # walk through each child directory in the designated local data folder
+        for root, directories, files in os.walk(local_data_path):
+            # update destination path if we have walked into a child directory
+            if root is not local_data_path:
+                # get the child directory relative path
+                subpath = os.path.relpath(root, local_data_path)
+                # update destination path to include child directories (ie subpaths)
+                dest_path = os.path.join(parent_dest_path, subpath)
+            # create child directories on endpoint
+                try:
+                    self.transfer_client.operation_mkdir(endpoint_id=endpoint_id, path=dest_path)
+                except TransferAPIError as e:
+                    raise IOError(f"Error while creating child directory {dest_path}: {e.message}") from e
+            # get local path to file to upload
+            for filename in files:
+                filepath = os.path.join(root, filename)
+                # upload file to destination path on endpoint
+                result = self._upload_file(filepath, https_base_url, dest_path, endpoint_id)
+                results.append(result)
+        return results
+
+    def _upload_file(self, filepath, https_base_url, dest_path, endpoint_id):
```

Comment:
Prime candidate for moving into a submodule: Not a function that people should use from the FoundryClient, but one that might get used elsewhere.